File: packages/background/native/Background.swift
import Foundation

/// CI stub that records scheduled background jobs (durable and in-process).
enum LucentBackground {
  private static let lock = NSLock()
  private static var jobs: [String: (version: Double, durable: Bool, payload: String)] = [:]
  private static var runCount: Double = 0

  static func schedule(
    jobId: String,
    payloadVersion: Double,
    payload: String,
    durable: Bool
  ) -> LucentBackgroundJobHandle {
    lock.lock()
    jobs[jobId] = (payloadVersion, durable, payload)
    lock.unlock()
    return LucentBackgroundJobHandle(
      jobId: jobId,
      payloadVersion: payloadVersion,
      durable: durable,
      payload: payload
    )
  }

  static func scheduledCount() -> Double {
    lock.lock()
    defer { lock.unlock() }
    return Double(jobs.count)
  }

  static func runCountValue() -> Double {
    lock.lock()
    defer { lock.unlock() }
    return runCount
  }

  static func remove(jobId: String) {
    lock.lock()
    jobs.removeValue(forKey: jobId)
    lock.unlock()
  }

  /// Run a scheduled job once. Rejects if `expectedVersion` does not match.
  static func runOnce(jobId: String, expectedVersion: Double) throws -> Double {
    lock.lock()
    defer { lock.unlock() }
    guard let job = jobs[jobId] else { throw LucentError(code: "NOT_SCHEDULED") }
    guard job.version == expectedVersion else { throw LucentError(code: "VERSION_MISMATCH") }
    runCount += 1
    return job.version
  }
}

final class LucentBackgroundJobHandle: @unchecked Sendable {
  private let lock = NSLock()
  private var isClosed = false
  private var cancelled = false
  let jobId: String
  let payloadVersion: Double
  let durable: Bool
  private let payload: String

  init(jobId: String, payloadVersion: Double, durable: Bool, payload: String) {
    self.jobId = jobId
    self.payloadVersion = payloadVersion
    self.durable = durable
    self.payload = payload
  }

  var closed: Bool {
    lock.lock()
    defer { lock.unlock() }
    return isClosed
  }

  func cancel() {
    lock.lock()
    cancelled = true
    lock.unlock()
    LucentBackground.remove(jobId: jobId)
  }

  /// Run this handle's payload once if still scheduled and version matches.
  func runOnce() throws -> Double {
    lock.lock()
    let cancelledNow = cancelled || isClosed
    let version = payloadVersion
    let id = jobId
    lock.unlock()
    guard !cancelledNow else { throw LucentError(code: "CANCELLED") }
    return try LucentBackground.runOnce(jobId: id, expectedVersion: version)
  }

  func close() async {
    await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
      lock.lock()
      isClosed = true
      lock.unlock()
      continuation.resume()
    }
  }
}

File: packages/background/native/Background.kt
package {{androidPackage}}

import kotlin.coroutines.Continuation
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

/** CI stub that records scheduled background jobs (durable and in-process). */
object LucentBackground {
  private val lock = Any()
  private val jobs = mutableMapOf<String, Triple<Double, Boolean, String>>()
  private var runCount = 0.0

  fun schedule(jobId: String, payloadVersion: Double, payload: String, durable: Boolean): LucentBackgroundJobHandle {
    synchronized(lock) { jobs[jobId] = Triple(payloadVersion, durable, payload) }
    return LucentBackgroundJobHandle(jobId, payloadVersion, durable, payload)
  }

  fun scheduledCount(): Double = synchronized(lock) { jobs.size.toDouble() }

  fun runCountValue(): Double = synchronized(lock) { runCount }

  fun remove(jobId: String) {
    synchronized(lock) { jobs.remove(jobId) }
  }

  /** Run a scheduled job once. Rejects if `expectedVersion` does not match. */
  fun runOnce(jobId: String, expectedVersion: Double): Double {
    synchronized(lock) {
      val job = jobs[jobId] ?: throw LucentError("NOT_SCHEDULED", "Job is not scheduled")
      if (job.first != expectedVersion) throw LucentError("VERSION_MISMATCH", "Payload version mismatch")
      runCount += 1.0
      return job.first
    }
  }
}

class LucentBackgroundJobHandle(
  val jobId: String,
  val payloadVersion: Double,
  val durable: Boolean,
  private val payload: String,
) {
  private val lock = Any()
  private var isClosed = false
  private var cancelled = false

  val closed: Boolean
    get() = synchronized(lock) { isClosed }

  fun cancel() {
    synchronized(lock) { cancelled = true }
    LucentBackground.remove(jobId)
  }

  /** Run this handle's payload once if still scheduled and version matches. */
  fun runOnce(): Double {
    val cancelledNow = synchronized(lock) { cancelled || isClosed }
    if (cancelledNow) throw LucentError("CANCELLED", "Job was cancelled")
    return LucentBackground.runOnce(jobId, payloadVersion)
  }

  suspend fun close(): Unit = suspendCoroutine { continuation ->
    synchronized(lock) { isClosed = true }
    continuation.resume(Unit)
  }
}
